demos/database.php

```diff
@@ -21,11 +21,40 @@ public function init()
     {
         parent::init();
         $this->addField('name', ['actual'=>'nicename', 'required'=>true, 'type'=>'string']);
+        $this->addField('sys_name', ['actual'=>'name', 'system'=>true]);
 
         $this->addField('iso', ['caption'=>'ISO', 'required'=>true, 'type'=>'string']);
         $this->addField('iso3', ['caption'=>'ISO3', 'required'=>true, 'type'=>'string']);
         $this->addField('numcode', ['caption'=>'ISO Numeric Code', 'type'=>'number', 'required'=>true]);
         $this->addField('phonecode', ['caption'=>'Phone Prefix', 'type'=>'number']);
+
+        $this->addHook('beforeSave', function ($m) {
+            if (!$m['sys_name']) {
+                $m['sys_name'] = strtoupper($m['name']);
+            }
+        });
+    }
+
+    public function validate()
+    {
+        $errors = parent::validate();
+
+        if (strlen($this['iso']) !== 2) {
+            $errors['iso'] = 'Must be exactly 2 characters';
+        }
+
+        if (strlen($this['iso3']) !== 3) {
+            $errors['iso3'] = 'Must be exactly 3 characters';
+        }
+
+        // look if name is unique
+        $c = clone $this;
+        $c->tryLoadBy('name', $this['name']);
```

Comment:
this still is not 100% safe way to check name uniqueness.
what is name already is not unique in database?